File: src/ast.rs
use crate::parser::CodeLocation;
use crate::value::ObjID;

// @NOTE:
// How zig does AST stuff with indexes instead of pointers / convential tree structure
// https://mitchellh.com/zig/parser
//

#[derive(Debug)]
pub struct AST {
	roots: Vec<usize>,
	nodes: Vec<Node>,
	extra_data: Vec<usize>,
	strings: Vec<String>,
	blocks: Vec<NodeBlock>,
	locations: Vec<CodeLocation>,
}

impl AST {
	pub fn new() -> Self {
		Self {
			roots: Vec::new(),
			nodes: Vec::new(),
			extra_data: Vec::new(),
			strings: Vec::new(),
			blocks: Vec::new(),
			locations: Vec::new(),
		}
	}

	pub fn add_node(&mut self, node: Node, location: CodeLocation) -> usize {
		self.nodes.push(node);
		self.locations.push(location);
		self.nodes.len() - 1
	}

	pub fn add_root(&mut self, root_index: usize) {
		self.roots.push(root_index);
	}

	pub fn add_extra_data<E>(&mut self, extra_data: E) -> usize
	where
		E: ExtraData,
	{
		let extra_data = extra_data.as_slice();
		let extra_data_index = self.extra_data.len();
		self.extra_data.extend_from_slice(extra_data);
		extra_data_index
	}

	pub fn add_string(&mut self, string: String) -> usize {
		if let Some(index) = self.strings.iter().position(|s| *s == string) {
			index
		} else {
			self.strings.push(string);
			self.strings.len() - 1
		}
	}

	pub fn add_block(&mut self, block: NodeBlock) -> usize {
		self.blocks.push(block);
		self.blocks.len() - 1
	}

	pub fn get(&self, index: usize) -> &Node {
		&self.nodes[index]
	}

	pub fn get_mut(&mut self, index: usize) -> &mut Node {
		&mut self.nodes[index]
	}

	pub fn get_extra_data<E>(&self, index: usize) -> E
	where
		E: ExtraData + Default + Sized,
	{
		let slice = self
			.extra_data
			.get(index..index + E::len())
			.expect("Failed to retrieve extra data!");

		E::from(slice)
	}
}

impl std::fmt::Display for AST {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		const INDENT_SIZE: usize = 2;

		fn fmt_at_indent(
			me: &AST,
			node_index: usize,
			indent: usize,
			f: &mut std::fmt::Formatter<'_>,
		) -> std::fmt::Result {
			let node = me.nodes[node_index];

			use NodeKind::*;
			match node.kind {
				// Literals
				Bool => write!(f, "Bool({})\n", node.lhs != 0),
				Int => write!(f, "Int({})\n", node.lhs as i64),
				Num => write!(f, "Num({})\n", unsafe {
					std::mem::transmute::<usize, f64>(node.lhs)
				}),
				Str => todo!(),
				Ident => write!(f, "Ident({})\n", me.strings[node.lhs]),
				List => todo!(),

				// Unary
				Negate => fmt_unary_at_indent("-", me, node_index, indent, f),

				// Binary
				Add => fmt_binary_at_indent("+", me, node_index, indent, f),
				Subtract => fmt_binary_at_indent("-", me, node_index, indent, f),
				Multiply => fmt_binary_at_indent("*", me, node_index, indent, f),
				Divide => fmt_binary_at_indent("/", me, node_index, indent, f),
				Assign => fmt_binary_at_indent("=", me, node_index, indent, f),

				// Blocks
				Block => fmt_block_at_indent("{}", me, node_index, indent, f),

				// Unique
				If => {
					let data: ExtraDataIf = me.get_extra_data(node.lhs);

					write!(f, "`if` {{\n")?;

					write!(
						f,
						"{: >indent$}condition: ",
						"",
						indent = (indent + 1) * INDENT_SIZE
					)?;
					fmt_at_indent(me, data.condition, indent + 1, f)?;

					write!(
						f,
						"{: >indent$}then: ",
						"",
						indent = (indent + 1) * INDENT_SIZE
					)?;
					fmt_at_indent(me, data.then_block, indent + 1, f)?;

					if data.else_block != 0 {
						write!(
							f,
							"{: >indent$}else: ",
							"",
							indent = (indent + 1) * INDENT_SIZE
						)?;
						fmt_at_indent(me, data.else_block, indent + 1, f)?;
					}

					write!(f, "{: >indent$}}}\n", "", indent = indent * INDENT_SIZE)
				}
				While => {
					write!(f, "`while` {{\n")?;

					write!(
						f,
						"{: >indent$}condition: ",
						"",
						indent = (indent + 1) * INDENT_SIZE
					)?;
					fmt_at_indent(me, node.lhs, indent + 1, f)?;

					write!(
						f,
						"{: >indent$}body: ",
						"",
						indent = (indent + 1) * INDENT_SIZE
					)?;
					fmt_at_indent(me, node.rhs, indent + 1, f)?;

					write!(f, "{: >indent$}}}\n", "", indent = indent * INDENT_SIZE)
				}
			}
		}

		fn fmt_unary_at_indent(
			op: &str,
			me: &AST,
			node_index: usize,
			indent: usize,
			f: &mut std::fmt::Formatter<'_>,
		) -> std::fmt::Result {
			let node = me.nodes[node_index];
			write!(f, "`{}` {{\n", op)?;
			write!(
				f,
				"{: >indent$}{}",
				"",
				"sub: ",
				indent = (indent + 1) * INDENT_SIZE
			)?;
			fmt_at_indent(me, node.lhs, indent + 1, f)?;
			write!(f, "{: >indent$}}}\n", "", indent = indent * INDENT_SIZE)
		}

		fn fmt_binary_at_indent(
			op: &str,
			me: &AST,
			node_index: usize,
			indent: usize,
			f: &mut std::fmt::Formatter<'_>,
		) -> std::fmt::Result {
			let node = me.nodes[node_index];
			write!(f, "`{}` {{\n", op)?;
			write!(
				f,
				"{: >indent$}{}",
				"",
				"lhs: ",
				indent = (indent + 1) * INDENT_SIZE
			)?;
			fmt_at_indent(me, node.lhs, indent + 1, f)?;
			write!(
				f,
				"{: >indent$}{}",
				"",
				"rhs: ",
				indent = (indent + 1) * INDENT_SIZE
			)?;
			fmt_at_indent(me, node.rhs, indent + 1, f)?;
			write!(f, "{: >indent$}}}\n", "", indent = indent * INDENT_SIZE)
		}

		fn fmt_block_at_indent(
			op: &str,
			me: &AST,
			node_index: usize,
			indent: usize,
			f: &mut std::fmt::Formatter<'_>,
		) -> std::fmt::Result {
			let node = me.nodes[node_index];
			let block = &me.blocks[node.lhs];
			write!(f, "`{}` {{\n", op)?;
			for (i, &root) in block.roots.iter().enumerate() {
				write!(
					f,
					"{: >indent$}{}. ",
					"",
					i,
					indent = (indent + 1) * INDENT_SIZE
				)?;
				fmt_at_indent(me, root, indent + 1, f)?;
			}
			write!(f, "{: >indent$}}}\n", "", indent = indent * INDENT_SIZE)
		}

		write!(f, "AST {{\n")?;
		for (i, &root) in self.roots.iter().enumerate() {
			write!(f, "{: >indent$}{}. ", "", i, indent = INDENT_SIZE)?;
			fmt_at_indent(self, root, 1, f)?;
		}
		write!(f, "}}")
	}
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum NodeKind {
	// Literals
	Bool,
	Int,
	Num,
	Str,
	Ident,
	List,

	// Unary
	Negate,

	// Binary
	Add,
	Subtract,
	Multiply,
	Divide,
	Assign,

	// Blocks
	Block,

	// Unique
	If,
	While,
}

#[derive(Clone, Copy, Debug)]
pub struct Node {
	pub kind: NodeKind,
	pub lhs: usize,
	pub rhs: usize,
}

impl Node {
	pub fn new(kind: NodeKind, lhs: usize, rhs: usize) -> Self {
		Self { kind, lhs, rhs }
	}

	pub fn new_bool(value: bool) -> Self {
		Self::new(NodeKind::Bool, value as usize, 0)
	}

	pub fn new_int(value: i64) -> Self {
		Self::new(NodeKind::Int, value as usize, 0)
	}

	pub fn new_num(value: f64) -> Self {
		Self::new(NodeKind::Num, unsafe { std::mem::transmute(value) }, 0)
	}

	pub fn new_str(index: usize) -> Self {
		Self::new(NodeKind::Str, index, 0)
	}

	pub fn new_ident(index: usize) -> Self {
		Self::new(NodeKind::Ident, index, 0)
	}

	pub fn new_list(value: ObjID) -> Self {
		Self::new(NodeKind::List, value, 0)
	}
}

#[derive(Debug)]
pub struct NodeBlock {
	pub kind: NodeKind,
	pub roots: Vec<usize>,
}

impl NodeBlock {
	pub fn new(kind: NodeKind, roots: Vec<usize>) -> Self {
		Self { kind, roots }
	}
}

pub trait ExtraData
where
	Self: Sized + Default,
{
	fn len() -> usize {
		std::mem::size_of::<Self>() / std::mem::size_of::<usize>()
	}

	fn from(slice: &[usize]) -> Self {
		assert_eq!(slice.len(), Self::len());
		let mut s = Self::default();
		s.as_slice_mut().copy_from_slice(slice);
		s
	}

	fn as_slice(&self) -> &[usize] {
		unsafe { std::slice::from_raw_parts(self as *const Self as *const usize, Self::len()) }
	}

	fn as_slice_mut(&mut self) -> &mut [usize] {
		unsafe { std::slice::from_raw_parts_mut(self as *mut Self as *mut usize, Self::len()) }
	}
}

impl ExtraData for ExtraDataIf {}

#[repr(C)] // We need repr(C) to guarentee fields are in the expencted places
#[derive(Default)]
pub struct ExtraDataIf {
	pub condition: usize,
	pub then_block: usize,
	pub else_block: usize, // 0 means no else_block
}

File: src/ast.zig
const std = @import("std");
const parser = @import("parser.zig");
const Token = parser.Token;

pub const Ast = struct {
    kind: AstKind,
    token: Token,

    const This = @This();

    fn init(kind: AstKind, token: Token) This {
        return This{ .kind = kind, .token = token };
    }

    pub fn downcast(this: *This, comptime T: type) *T {
        return @ptrCast(*T, this);
    }

    pub fn downcastConst(this: *const This, comptime T: type) *const T {
        return @ptrCast(*const T, this);
    }

    pub fn format(this: *const This, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) @TypeOf(writer).Error!void {
        switch (this.kind) {
            // Literals
            .Bool, .Int, .Num, .Str => try writer.print("{}", .{this.downcastConst(AstLiteral)}),
            .Ident => try writer.print("{}", .{this.downcastConst(AstIdent)}),

            // Unary
            .Negate => try writer.print("{}", .{this.downcastConst(AstUnary)}),

            // Binary
            .Assign, .Add, .Subtract, .Multiply, .Divide => try writer.print("{}", .{this.downcastConst(AstBinary)}),

            // Blocks
            .Block, .Comma => try writer.print("{}", .{this.downcastConst(AstBlock)}),

            .If => try writer.print("{}", .{this.downcastConst(AstIf)}),
            .While => try writer.print("{}", .{this.downcastConst(AstWhile)}),
            .Def => try writer.print("{}", .{this.downcastConst(AstDef)}),
            .Var => try writer.print("{}", .{this.downcastConst(AstVar)}),
        }
    }
};

pub const AstKind = enum {
    // Literals
    Bool,
    Int,
    Num,
    Str,
    Ident,

    // Unary
    Negate,

    // Binary
    Assign,
    Add,
    Subtract,
    Multiply,
    Divide,

    // Blocks
    Block,
    Comma,

    If,
    While,
    Def,
    Var,
};

pub const AstLiteral = struct {
    kind: AstKind,
    token: Token,
    literal: Literal,

    const This = @This();

    pub const Literal = union(enum) {
        Bool: bool,
        Int: i64,
        Num: f64,
        Str: []const u8,
    };

    pub fn init(kind: AstKind, token: Token, literal: Literal) This {
        return This{ .kind = kind, .token = token, .literal = literal };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstIdent = struct {
    kind: AstKind,
    token: Token,
    ident: []const u8,

    const This = @This();

    pub fn init(token: Token, ident: []const u8) This {
        return This{ .kind = .Ident, .token = token, .ident = ident };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstUnary = struct {
    kind: AstKind,
    token: Token,
    sub: *Ast,

    const This = @This();

    pub fn init(kind: AstKind, token: Token, sub: *Ast) This {
        return This{ .kind = kind, .token = token, .sub = sub };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstBinary = struct {
    kind: AstKind,
    token: Token,
    lhs: *Ast,
    rhs: *Ast,

    const This = @This();

    pub fn init(kind: AstKind, token: Token, lhs: *Ast, rhs: *Ast) This {
        return This{ .kind = kind, .token = token, .lhs = lhs, .rhs = rhs };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstBlock = struct {
    kind: AstKind,
    token: Token,
    nodes: []*Ast,

    const This = @This();

    pub fn init(kind: AstKind, token: Token, nodes: []*Ast) This {
        return This{ .kind = kind, .token = token, .nodes = nodes };
    }

    pub fn deinit(this: *This, allocator: std.mem.Allocator) void {
        allocator.free(this.nodes);
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstIf = struct {
    kind: AstKind,
    token: Token,
    condition: *Ast,
    then_block: *AstBlock,
    else_block: ?*Ast,

    const This = @This();

    pub fn init(kind: AstKind, token: Token, condition: *Ast, then_block: *AstBlock, else_block: ?*Ast) This {
        return This{ .kind = kind, .token = token, .condition = condition, .then_block = then_block, .else_block = else_block };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstWhile = struct {
    kind: AstKind,
    token: Token,
    condition: *Ast,
    block: *AstBlock,

    const This = @This();

    pub fn init(kind: AstKind, token: Token, condition: *Ast, block: *AstBlock) This {
        return This{ .kind = kind, .token = token, .condition = condition, .block = block };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstDef = struct {
    kind: AstKind,
    token: Token,
    name: []const u8,
    params: *AstBlock,
    body: *AstBlock,

    const This = @This();

    pub fn init(token: Token, name: []const u8, params: *AstBlock, body: *AstBlock) This {
        return This{ .kind = .Def, .token = token, .name = name, .params = params, .body = body };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};

pub const AstVar = struct {
    kind: AstKind,
    token: Token,
    ident: *AstIdent,
    initializer: *Ast,

    const This = @This();

    pub fn init(token: Token, ident: *AstIdent, initializer: *Ast) This {
        return This{ .kind = .Var, .token = token, .ident = ident, .initializer = initializer };
    }

    pub fn asAst(this: *This) *Ast {
        return @ptrCast(*Ast, this);
    }
};
